File: python/shapes.py
from p5 import *


class Row:
    def __init__(self, y_pos=1, x_positions=[1, 2], grid_width=10, color=255):
        self.y_pos = y_pos
        self.x_positions = x_positions
        self.grid_width = grid_width
        self.color = color
        self.circles = [
            Circle(p, self.y_pos * self.grid_width, color=self.color)
            for p in self.x_positions
        ]

    def draw(self):
        for circle in self.circles:
            circle.draw()

    def get_circles(self):
        return self.circles


class Circle:
    # a circle just contains its x, y position, colour and radius.
    def __init__(self, x, y, radius=7, color=50):
        self.x = x
        self.y = y
        self.radius = radius
        self.color = color

    def is_clicked(self, x, y):
        distance = sqrt((x - self.x) ** 2 + (y - self.y) ** 2)
        return distance <= self.radius

    def change_color(self, color):
        self.color = color

    def draw(self):
        # Unpack the color tuple here
        if isinstance(self.color, tuple):
            fill(*self.color)
        else:
            fill(self.color)
        circle((self.x, self.y), self.radius)

File: app/simulation/shapes.ts
import { P5CanvasInstance } from "@p5-wrapper/react";
import { Color } from "./types";

export class Circle {
  constructor(
    private x: number,
    private y: number,
    private radius = 7,
    private color: Color | number = 50
  ) {}

  public isClicked(x: number, y: number) {
    const distance = Math.sqrt((x - this.x) ** 2 + (y - this.y) ** 2);
    return distance <= this.radius;
  }

  public changeColor(color: number | Color) {
    this.color = color;
  }

  public draw(p5: P5CanvasInstance) {
    if (Array.isArray(this.color)) {
      p5.fill(this.color[0], this.color[1], this.color[2]);
    } else {
      p5.fill(this.color);
    }

    // console.log(this.x, this.y, this.radius);
    p5.circle(this.x, this.y, this.radius);
  }
}

export class Row {
  constructor(
    private yPosition = 1,
    private xPositions = [1, 2],
    private gridWidth = 10,
    private color: number | Color = 255,
    private circles: Circle[] = []
  ) {
    this.circles = xPositions.map(
      (xPosition) =>
        new Circle(
          xPosition,
          this.yPosition * this.gridWidth,
          undefined,
          this.color
        )
    );
  }

  public draw(p5: P5CanvasInstance) {
    this.circles.forEach((circle) => {
      circle.draw(p5);
    });
  }

  public getCircles() {
    return this.circles;
  }
}
